### PR TITLE
Userlog Service

### PR DESCRIPTION
**Pull request by kobergj**
Backport number 2 from https://github.com/owncloud/ocis/pull/5506

Introduces the userlog service

### CODE REVIEW

File: services/userlog/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+The `userlog` service provides a way to configure which events a user wants to be informed about and an API to retrieve them.
+
+## Configuring
+
+The `userlog` service has hardcoded configuration for now. 
+
+## Retrieving
+
+The `userlog` service provides an API to retrieve configured events.
```
Comment (mmattel) [inline services/userlog/README.md:11]:
```suggestion
# Userlog Service

The `userlog` service is a mediator between the `eventhistory` service and clients who want to be informed about user related events. It provides an API to retrieve those events.

## Configuring

For the time being, the configuration is hardcoded and cannot be changed.

## Retrieving

The `userlog` service provides an API to retrieve configured events. 
```

File: services/userlog/pkg/config/config.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	Events            Events `yaml:"events"`
+	Store             Store  `yaml:"store"`
+
+	Context context.Context `yaml:"-"`
+}
+
+// Store configures the store to use
+type Store struct {
+	Type         string        `yaml:"type" env:"USERLOG_STORE_TYPE" desc:"The type of the store. Supported is inmemory"`
+	RecordExpiry time.Duration `yaml:"record_expiry" env:"USERLOG_RECORD_EXPIRY" desc:"time to life for events in the store"`
```
Comment (mmattel) [inline services/userlog/pkg/config/config.go:33]:
```suggestion
	RecordExpiry time.Duration `yaml:"record_expiry" env:"USERLOG_RECORD_EXPIRY" desc:"Time to life for events in the store. The duration can be set as number followed by a unit identifier like s, m or h."`
```
Maybe to add `Defaults to '336h' (2 weeks).` (change the number)

File: services/userlog/pkg/config/config.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	RevaGateway       string `yaml:"reva_gateway" env:"REVA_GATEWAY" desc:"CS3 gateway used to look up user metadata"`
+	Events            Events `yaml:"events"`
+	Store             Store  `yaml:"store"`
+
+	Context context.Context `yaml:"-"`
+}
+
+// Store configures the store to use
+type Store struct {
+	Type         string        `yaml:"type" env:"USERLOG_STORE_TYPE" desc:"The type of the store. Supported is inmemory"`
```
Comment (mmattel) [inline services/userlog/pkg/config/config.go:32]:
```suggestion
	Type         string        `yaml:"type" env:"USERLOG_STORE_TYPE" desc:"The type of the userlog store. Supported values are: 'mem'. See the text description for details."`
```

File: services/userlog/pkg/config/config.go (hunk 2 above)
Comment (kobergj) [reply services/userlog/pkg/config/config.go]: No expiry for userlog service

File: services/userlog/pkg/config/config.go (hunk 3 above)
Comment (kobergj) [reply services/userlog/pkg/config/config.go]: Copied descriptions from eventhistory service

File: services/userlog/README.md (hunk 1 above)
Comment (kobergj) [reply services/userlog/README.md]: Updated readme. Please recheck

File: services/userlog/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+1.  Note that in-memory stores are by nature not reboot persistent.
+2.  Though usually not necessary, a database name and a database table can be configured for event stores if the event store supports this. Generally not applicapable for stores of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store applies.
+
+## Configuring
+
+For the time being, the configuration which user related events are of interest is hardcoded and cannot be changed.
+
+## Retrieving
+
+The `userlog` service provides an API to retrieve configured events. For now this API is mostly following [oc10 notification GET API](https://doc.owncloud.com/server/next/developer_manual/core/apis/ocs-notification-endpoint-v1.html#get-user-notifications)
```
Comment (mmattel) [inline services/userlog/README.md:24]:
```suggestion
The `userlog` service provides an API to retrieve configured events. For now, this API is mostly following the [oc10 notification GET API](https://doc.owncloud.com/server/next/developer_manual/core/apis/ocs-notification-endpoint-v1.html#get-user-notifications).
```

File: services/userlog/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+The `userlog` service persists information via the configured store in `USERLOG_STORE_TYPE`. Possible stores are:
+  -   `mem`: Basic in-memory store and the default.
+  -   `ocmem`: Advanced in-memory store allowing max size.
+  -   `redis`: Stores data in a configured redis cluster.
+  -   `etcd`: Stores data in a configured etcd cluster.
+  -   `nats-js`: Stores data using key-value-store feature of [nats jetstream](https://docs.nats.io/nats-concepts/jetstream/key-value-store)
+  -   `noop`: Stores nothing. Useful for testing. Not recommended in productive enviroments.
+
+1.  Note that in-memory stores are by nature not reboot persistent.
+2.  Though usually not necessary, a database name and a database table can be configured for event stores if the event store supports this. Generally not applicapable for stores of type `in-memory`. These settings are blank by default which means that the standard settings of the configured store applies.
```
[the remark this answers is not in the document]
Comment (mmattel) [reply services/userlog/README.md:16]: done, see #5619

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>wopiValidatorTests-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/19351/42/1) failed. Further test are cancelled...

File: services/userlog/README.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
+# Userlog Service
+
+The `userlog` service is a mediator between the `eventhistory` service and clients who want to be informed about user related events. It provides an API to retrieve those.
+
+## Store
```
Comment (mmattel) [inline services/userlog/README.md:5]:
```suggestion
## Prerequisites

Running the `userlog` service without running the `eventhistory` service is not possible.

## Store
```

File: services/userlog/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+For the time being, the configuration which user related events are of interest is hardcoded and cannot be changed.
+
+## Retrieving
+
+The `userlog` service provides an API to retrieve configured events. For now, this API is mostly following the [oc10 notification GET API](https://doc.owncloud.com/server/next/developer_manual/core/apis/ocs-notification-endpoint-v1.html#get-user-notifications).
+
+## Deleting
+
+To delete events for an user use a `DELETE` request to `ocs/v2.php/apps/notifications/api/v1/notifications` containing the ids to delete.
```
Comment (mmattel) [inline services/userlog/README.md:29]:
```suggestion
To delete events for an user, use a `DELETE` request to `ocs/v2.php/apps/notifications/api/v1/notifications` containing the ID's to delete.
```

**Comment by mmattel** [comment]
The texts and descriptions are ok from my side.

**Review by C0rby — APPROVED** [review]

**Pull request merged**